injecting function details into TraceEvent.Properties

brettsam opened the pull request:
For #774 and #784.

@mathewc -- The tests are not ready yet but wanted a quick review of the approach here. This adds the hostId, instanceId, and functionDescriptor values to the Properties collection in both our `TraceWriterFunctionInstanceLogger` and the `TraceWriter` we pass to the function itself.

I investigated including the trigger object as well. I could inject the *object* fairly easily with the `TraceWriter` that we pass as a parameter to functions. And I can include the *string* from `ToInvokeString()` with the `TraceWriterFunctionInstanceLogger`, but that felt odd to have one form in one message and another elsewhere. I'll keep looking into that.

brettsam commented:
Ready for review.

mathewc commented on src/Microsoft.Azure.WebJobs.Host/Loggers/TraceEventExtensions.cs:
```diff
@@ -0,0 +1,22 @@
+        public static void AddFunctionInstanceDetails(this TraceEvent traceEvent, Guid hostId, FunctionDescriptor descriptor, Guid functionId)
+        {
+            traceEvent.Properties[HostIdKey] = hostId;
+            traceEvent.Properties[FunctionDescriptorKey] = descriptor;
```
FunctionDescriptor is public, right? So you're doing this as a way to include the function Name (which was the original ask) as well as additional details?

brettsam replied:
Yeah it's public. It includes a few other things that someone may want -- I didn't have a good reason to restrict them: https://github.com/Azure/azure-webjobs-sdk/blob/master/src/Microsoft.Azure.WebJobs.Protocols/FunctionDescriptor.cs

mathewc commented on src/Microsoft.Azure.WebJobs.Host/Loggers/TraceEventExtensions.cs:
```diff
@@ -0,0 +1,22 @@
+
+        public static void AddFunctionInstanceDetails(this TraceEvent traceEvent, Guid hostId, FunctionDescriptor descriptor, Guid functionId)
+        {
+            traceEvent.Properties[HostIdKey] = hostId;
```
Wondering what the scenario is for HostId? This will be the same for all logs, right? So I'm wondering how this would be useful for someones custom trace writer. If they wanted HostId they could access that themselves by passing it into their custom tracer on construction.

brettsam replied:
HostID was explicitly asked for in #774. Don't you have to construct and register your tracer with the Config before you have a Host id? That's what I thought -- but I can double-check if there's any way to do that.

mathewc replied:
They were asking for HostInstanceId not HostId. HostId is the same across all scaled out instances, HostInstanceId is unique per instance.

brettsam replied:
Well, I learned something today :-)

mathewc commented on src/Microsoft.Azure.WebJobs.Host/Loggers/TraceWriterFunctionInstanceLogger.cs:
```diff
@@ -24,28 +25,38 @@ public TraceWriterFunctionInstanceLogger(TraceWriter trace)
         {
-            _trace.Info(string.Format(CultureInfo.InvariantCulture, "Executing: '{0}' - Reason: '{1}'", message.Function.ShortName, message.FormatReason()), TraceSource.Execution);
+            string traceMessage = string.Format(CultureInfo.InvariantCulture, "Executing: '{0}' - Reason: '{1}'", message.Function.ShortName, message.FormatReason());
+            Trace(message.HostInstanceId, message.Function, message.FunctionInstanceId, traceMessage, TraceLevel.Info, TraceSource.Execution);
```
nit: move the TraceLevel param to the front for clarity

mathewc commented on src/Microsoft.Azure.WebJobs.Host/Executors/FunctionInstanceTraceWriter.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace Microsoft.Azure.WebJobs.Host.Executors
+{
+    internal class FunctionInstanceTraceWriter : TraceWriter
```
nit: add a comment block to this guy describing what he's for

mathewc commented:
Agree with our general approach. I'd forget about adding the trigger object or anything like that for now.

brettsam commented:
Well, turns out I was using `HostInstanceId` anyway. I just was using the wrong parameter names -- I've updated all of that. Addressed other comments as well. Anything else?

mathewc commented:
:ship:
